### PR TITLE
crimson/seastore: add lba tree metrics for lba tree profiling

### PR DESCRIPTION
**Pull request by liu-chunmei**
Signed-off-by: chunmei-liu <chunmei.liu@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/os/seastore/cache.cc [hunk 1]
```diff
@@ -501,6 +501,52 @@ void Cache::register_metrics()
...
+        ),
+        sm::make_counter(
+          "lba_tree_iter_next_invalidated",
+          invalidated_lba_tree_efforts.num_iter_nexts,
+          sm::description("total number of invalidated lba iter_next operations"),
+          {tree_label}
+        ),
+      }
+    );
+  };
```
Comment (cyx1231st) [inline src/crimson/os/seastore/cache.cc:549]: I think the goal is to measure the number of iterations vs the number of allocations during laddr-alloc, but I don't think it worths to relate to transaction invalidate/commit in `Cache`.

So could you define these metrics inside the lba tree instead?

File: src/crimson/os/seastore/lba_manager/btree/lba_btree.cc [hunk 2]
```diff
@@ -28,6 +28,7 @@ LBABtree::iterator_fut LBABtree::iterator::next(
   assert_valid();
   assert(!is_end());
 
+  ++(c.trans.get_lba_tree_inner_stats().num_iter_nexts);
```
Comment (cyx1231st) [inline src/crimson/os/seastore/lba_manager/btree/lba_btree.cc:31]: There are other operations calls `next()` and `prev()`, so the result is not accurate by dividing num_iter_nexts by num_alloc_extents.

Comment (liu-chunmei) [reply src/crimson/os/seastore/lba_manager/btree/lba_btree.cc:31]: yes, get_mappings ,scan_mappings and scan_mapped_space also can cause iter.next is called, will correct it.

File: src/crimson/os/seastore/cache.cc (hunk 1 above)
Comment (liu-chunmei) [reply src/crimson/os/seastore/cache.cc:549]: sure, will do it.

File: src/crimson/os/seastore/lba_manager/btree/lba_btree.cc (hunk 2 above)
Comment (athanatos) [reply src/crimson/os/seastore/lba_manager/btree/lba_btree.cc:31]: Yeah, that's what I was getting at.  Sounds good.

**Comment by liu-chunmei** [comment]
with the lba metircs register in LBABtreeManager,  run /cbt/tools/crimson/seastore_fio_run.sh iter.next is still much bigger than alloc_extent. FYI.
before run fio:
    "LBA_lba_tree_alloc_extents": {
        "shard": "0",
        "tree": "LBA",
        "value": 312
    },
    "LBA_lba_tree_iter_nexts": {
        "shard": "0",
        "tree": "LBA",
        "value": 271
    },
    "LBA_lba_tree_iter_prevs": {
        "shard": "0",
        "tree": "LBA",
        "value": 286
    },
run fio 600s:
    "LBA_lba_tree_alloc_extents": {
        "shard": "0",
        "tree": "LBA",
        "value": 756268
    },
    "LBA_lba_tree_iter_nexts": {
        "shard": "0",
        "tree": "LBA",
        "value": 1375180973
    },
    "LBA_lba_tree_iter_prevs": {
        "shard": "0",
        "tree": "LBA",
        "value": 654613
    },
run another 600s:
    "LBA_lba_tree_alloc_extents": {
        "shard": "0",
        "tree": "LBA",
        "value": 1050906
    },
    "LBA_lba_tree_iter_nexts": {
        "shard": "0",
        "tree": "LBA",
        "value": 3244344902
    },
    "LBA_lba_tree_iter_prevs": {
        "shard": "0",
        "tree": "LBA",
        "value": 859811
    },
run third 600s
    "LBA_lba_tree_alloc_extents": {
        "shard": "0",
        "tree": "LBA",
        "value": 1354641
    },
    "LBA_lba_tree_iter_nexts": {
        "shard": "0",
        "tree": "LBA",
        "value": 5422249397
    },
    "LBA_lba_tree_iter_prevs": {
        "shard": "0",
        "tree": "LBA",
        "value": 1067228
    },

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc [hunk 3]
```diff
@@ -424,7 +429,41 @@ BtreeLBAManager::BtreeLBAManager(
...
+void BtreeLBAManager::register_metrics()
+{
+  namespace sm = seastar::metrics;
+  auto tree_label = sm::label("tree");
+  auto lba_label = tree_label("LBA");
+  metrics.add_group(
+    "LBA",
+    {
+      sm::make_counter(
+        "lba_tree_alloc_extents",
```
Comment (cyx1231st) [inline src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc:447]: The final metric name will look like "LBA_lba_tree_alloc_extents" with label "tree=LBA".

So I think we can remove "lba_tree_" from the metric names, also remove the "tree=LBA" label as they are all duplicated information.

File: src/crimson/os/seastore/lba_manager/btree/lba_btree.h [hunk 4]
```diff
@@ -212,16 +218,20 @@ class LBABtree {
...
       c, addr
-    ).si_then([c, addr](auto iter) {
+    ).si_then([c, addr, need_count](auto iter) {
       if (iter.is_begin()) {
 	return iterator_fut(
 	  interruptible::ready_future_marker{},
 	  iter);
       } else {
+	if (need_count) {
+	  ++LBABtree::lba_tree_inner_stats.num_iter_prevs;
```
Comment (cyx1231st) [inline src/crimson/os/seastore/lba_manager/btree/lba_btree.h:233]: I think this metric doesn't provide very useful information for the goal, as we can directly see from the code that, if no transaction conflict happens, each `allocate_extent()` will increment this metric only once.

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc [hunk 5]
```diff
@@ -424,7 +429,41 @@ BtreeLBAManager::BtreeLBAManager(
...
+        {lba_label}
+      ),
+      sm::make_counter(
+        "lba_tree_iter_prevs",
+        LBABtree::lba_tree_inner_stats.num_iter_prevs,
+        sm::description("total number of lba iter_prev operations"),
+        {lba_label}
+      ),
+      sm::make_counter(
+        "lba_tree_iter_nexts",
```
Comment (cyx1231st) [inline src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc:459]: It's more accurate to be called, for example, "alloc_extents_iter_nexts", with the description "total number of iterator next operations during extent allocation"

**Comment by liu-chunmei** [comment]
@cyx1231st updated.

**Review by cyx1231st — APPROVED** [review]

File: src/crimson/os/seastore/lba_manager/btree/lba_btree.h [hunk 6]
```diff
@@ -27,6 +27,11 @@ class LBABtree {
 
   using mapped_space_visitor_t = LBAManager::scan_mapped_space_func_t;
 
+  struct lba_tree_inner_stats_t {
+    uint64_t num_alloc_extents = 0;
+    uint64_t num_iter_nexts = 0;
```
Comment (athanatos) [inline src/crimson/os/seastore/lba_manager/btree/lba_btree.h:32]: Rename this to num_alloc_extents_iter_nexts

**Review by athanatos — APPROVED** [review]
LGTM except for the variable rename above.

**Pull request merged**